[lamchenghou] iP

lamchenghou opened the pull request:
# Dot
> See your tasks as a constellation of stars, and dot them off one-by-one! 🤩

## Dot's features v.s. Duke
|Dot|Duke|
| :---: | :----: |
|Strong error handling|Minimal error handling|
|Both query by date & text|Query by text only|
|Funny personality|A personality a lot of chatbots have 🤔|

## User Experience & Installation
To improve your experience, on top of **iP** requirements, we've added:
- a help menu accessible using the `help` command
- robust error handling, it can even detect specific missing (or too many) input arguments!
- `whatsgoingon <dd/MM/yyyy>` to query the day's deadlines or events

<br />

**_Here's how you can get Dot!_**
1. download the latest release [here](https://github.com/lamchenghou/ip/releases/tag/Level-10)
1. in terminal, change your directory to the one that contains `dot.jar`
1. run `java -jar dot.jar`
1. sit back and have your tasks managed for you!  

## For developers
What do you think was the `regex` using to check against our data argument in the command `whatsgoingon <dd/MM/yyyy>`?
- [ ] `^([1-9]|0[1-9]|[1-2][0-9]|3[0-1])/([1-9]|0[1-9]|1[0-2])/[0-9]{4}$`
- [ ] `^(0[1-9]|[1-2][0-9]|3[0-1])/([1-9]|0[1-9]|1[0-2])/[0-9]{4}$`
- [ ] `^(0[1-9]|[1-2][0-9]|3[0-1])([1-9]|0[1-9]|1[0-2])[0-9]{4}$`
<details>
<summary> Answer</summary>
The <code>regex</code> used was <code>^([1-9]|0[1-9]|[1-2][0-9]|3[0-1])/([1-9]|0[1-9]|1[0-2])/[0-9]{4}$</code>. 
<br />Examples of dates we can support are <code>2/3/2017</code> and <code>06/12/2040</code>.
</details>

Here is a short snippet of the use of regex for validation in `isValidDate` of `parser` package:
```java
public static boolean isValidDate(String dateInput) {
    String dateRegex = "^([1-9]|0[1-9]|[1-2][0-9]|3[0-1])/([1-9]|0[1-9]|1[0-2])/[0-9]{4}$";
    return dateInput.matches(dateRegex);
}
```

## Incoming features
> [!NOTE]
> Check back weekly for new releases, up till the end of iP!
- [x] GUI

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

lamchenghou committed:
Add greeting messages

lamchenghou committed:
Reduce complexity of greeting message for first increment

lamchenghou committed:
Add echo functionality

lamchenghou committed:
Add dots personality

lamchenghou committed:
Add ui package and Ui.java

lamchenghou committed:
Add Task class

lamchenghou committed:
Add TaskList class

lamchenghou committed:
Add insert item and list functionality

lamchenghou committed:
Add wrapPrintWithHorizontalRules to ui package

lamchenghou committed:
Add maxSize to TaskList, which can vary from instance to instance

lamchenghou committed:
Fix scanner not reading whole line

lamchenghou committed:
Change tasks from ArrayList of String to that of Tasks

lamchenghou committed:
Add displayMarkOrUnmark Ui method

lamchenghou committed:
Add status toggle for Task and TaskList, and support new Ui for status

lamchenghou committed:
Add TaskErrors enum

lamchenghou committed:
Add IncorrectMarkParametersException

lamchenghou committed:
Add mark function to console input switch

lamchenghou committed:
Add ERR_USING_UNMARK to TaskErrors enum

lamchenghou committed:
Add unmark functionality to Dot

lamchenghou committed:
Add Todo, Deadline, Event subclasses of Task

lamchenghou committed:
Make Task and its subclasses public as it is used in Dot

lamchenghou committed:
Integrate todo, event, deadline commands into Dot

lamchenghou committed:
Update task message shown upon toggling status

lamchenghou committed:
Update regression testing files

lamchenghou committed:
Add more personality and variation to existing mark and unmark error handling

lamchenghou committed:
Add DotException to represent exceptions specific to Dot

lamchenghou committed:
Add more errors to TaskErrors enum, and improve error message for NumberFormatException

lamchenghou committed:
Touch up and add error handling for commands in Dot

lamchenghou committed:
Add error for deleting task into enum

lamchenghou committed:
Add delete functionality to TaskList

lamchenghou committed:
Add delete command to Dot

lamchenghou committed:
Update regression testing

lamchenghou committed:
Change from using RuntimeException to checked exception for DotException, but retain support for RunTimeExceptions like NumberFormatException

lamchenghou committed:
Fix typo in error name and update test files to include delete

lamchenghou committed:
Add save functionality, starting with dot.txt

lamchenghou committed:
Add overloading constructor to task subclasses to take in isCompleted, and also text file line format for each

lamchenghou committed:
Add new errors to TaskError enum

lamchenghou committed:
Add load and save functionality to Dot

lamchenghou committed:
Refactor by adding path name as const

lamchenghou committed:
Merge branch 'branch-Level-7'

lamchenghou committed:
Add Dateable optional class

lamchenghou committed:
Integrate Dateable into Deadline and Event

lamchenghou committed:
Fix event command parsing bug

lamchenghou committed:
Add length to Ui HORIZONTAL RULE

lamchenghou committed:
Add comparison methods to Dateable

lamchenghou committed:
Add error for cmd whatsgoingon to TaskError enum

lamchenghou committed:
Add parser functions for datetime

lamchenghou committed:
Add validation method for date

lamchenghou committed:
Add isOnDate method to all Task classes

We want to be able to pass in a LocalDateTime to a Task
and be able to tell if the Task has a Date that falls
within the queried Date.

This is needed to support the date query function to
find what deadlines and events fall within the queried date.

Let's apply polymorphism such that all subclasses of Task
override the comparison method in Task.

lamchenghou committed:
Add whatsgoingon dd/MM/yyyy command to Dot

lamchenghou committed:
Revert dot.txt to blank state

lamchenghou committed:
Merge branch 'branch-Level-8'

lamchenghou committed:
Fix date storage bug

lamchenghou committed:
Add help command to show help menu

lamchenghou committed:
Change Dot to nonstatic

lamchenghou committed:
Move validation and parsing functionalities to Validation.java and Parser.java

lamchenghou committed:
Fix condition for Event::isOnDate

lamchenghou committed:
Move Validation.java to parser package

lamchenghou committed:
Add commands package

lamchenghou committed:
Add parsing input to Command to Parser

lamchenghou committed:
Integrate new command system to Dot

lamchenghou committed:
Move stdin readline to package ui

lamchenghou committed:
Fix Dateable::isBefore/AfterOrOn

lamchenghou committed:
Merge branch 'branch-A-MoreOOP'

lamchenghou committed:
Merge branch 'add-gradle-support'

lamchenghou committed:
Add checkstyle configuration files

lamchenghou committed:
Modify build.gradle to add checkstyle and Dot pathname

lamchenghou committed:
refactor: Add Ui function to format print ArrayList of String

lamchenghou committed:
Merge branch 'branch-A-Gradle'

lamchenghou committed:
Add Parser JUnit test cases

lamchenghou committed:
Add Validation JUnit test cases

lamchenghou committed:
Add Event JUnit test cases

lamchenghou committed:
Add TaskList JUnit test cases

lamchenghou committed:
Update text ui testing script.

Dot has many changes, and will need more up-to-date testing.

Let's create a separate data file for testing, and the script
will access Dot in testing mode.

This is to allow Dot to be tested and not lose current data.

lamchenghou committed:
Merge branch 'branch-A-JUnit'

lamchenghou committed:
Update file structure, build.gradle, .gitignore, runtest.sh

lamchenghou committed:
Refactor codebase to allow Storage to become nonstatic

Initially the data text file was hardcoded in many locations.
We want to encapsulate that into the Storage class.

Let's make Storage nonstatic, reducing the need to pass
around dataFilePathname. Let's also revise storage to
create a new ./data/dot.txt file if it doesn't exist.

This also reduces the amount of parameters, making the code
cleaner in general.

It is good to note that running the source code Dot.java
will do it for the project root ip. For text-ui-test,
it is actually executed in the directory text-ui-test.
For JUnit, its runs it in the project root.

lamchenghou committed:
Add Storage changes to TaskList

lamchenghou committed:
Integrate Storage changes to Dot

lamchenghou committed:
Fix style in a portion of the codebase

lamchenghou committed:
Merge branch 'branch-A-Jar'

lamchenghou committed:
Add javadocs for all classes

lamchenghou committed:
Move all classes to dot package

lamchenghou committed:
Refactor codebase for move of files to dot package

lamchenghou committed:
Update runtest.sh to follow move to dot package

lamchenghou committed:
Update .gitignore

lamchenghou committed:
Update runtest.bat for Windows support

lamchenghou committed:
Update .gitignore

lamchenghou committed:
Modify boolean names to fit coding standard

lamchenghou committed:
Modify codebase to follow coding standard

lamchenghou committed:
Update javadocs to follow coding standard

lamchenghou committed:
Update help menu to include new Find command

lamchenghou committed:
Add find functionality to Task and TaskList

lamchenghou committed:
Add TaskError for find

lamchenghou committed:
Add FindCommand.java

lamchenghou committed:
Add FindCommand to parser

lamchenghou committed:
Update text-ui-test to test new feature

lamchenghou committed:
Merge branch 'branch-A-JavaDoc'

lamchenghou committed:
Merge branch 'branch-A-CodingStandard'

lamchenghou committed:
Fix merge issues

lamchenghou committed:
Merge branch 'branch-Level-9'

lamchenghou committed:
Update .gitignore

lamchenghou committed:
Untrack data directory

lamchenghou committed:
Enable checkstyle by editing build.gradle

lamchenghou committed:
Fix checkstyle errors in JUnit tests as detected by gradle

lamchenghou committed:
Merge branch 'branch-A-CheckStyle'

lamchenghou committed:
Setup gradle for JavaFX

lamchenghou committed:
Add MainWindow.fxml and MessageBox.fxml

lamchenghou committed:
Add user, dot and background image resources

lamchenghou committed:
Add MainWindowStyle.css

lamchenghou committed:
Update Ui to return String instead of printing

lamchenghou committed:
Add controllers for MainWindow and MessageBox

lamchenghou committed:
Add Main.java and Launcher.java as new entry point

lamchenghou committed:
Update build.gradle to follow new entry point

lamchenghou committed:
Update error handling workflow

lamchenghou committed:
Update commands to utilise GUI

lamchenghou committed:
Update Dot to follow GUI

lamchenghou committed:
Fix some style issues

lamchenghou committed:
Update Task and all other related classes for GUI

The move from text UI to GUI requires us to define
handlers that will update the GUI, instead of directly printing.

It is essential to refactor the whole codebase to follow
the GUI change.

Let's take all methods that print to the previous text UI
and turn them to methods that can receive a Consumer
which will update the GUI. We also make a handler capable
of handling errors, that is show the error in the GUI.

Passing GUI handlers around is easier than passing the
to-be-displayed output.

lamchenghou committed:
Update JUnit tests to follow GUI

lamchenghou committed:
Fix style issues

lamchenghou committed:
Merge branch 'branch-Level-10'

shuyangk commented on src/main/java/dot/commands/Command.java:
I like how you implemented the Commands!

shuyangk commented on src/main/java/dot/tasks/Dateable.java:
I like how you implemented this Dateable class to assist with reading date/time inputs!

shuyangk commented on src/main/java/dot/tasks/Task.java:
```diff
@@ -0,0 +1,102 @@
+
+    private final String description;
+
+    private boolean isCompleted;
```
I like how you named your boolean variable!

shuyangk commented on src/main/java/dot/tasks/TaskList.java:
I like how clearly you named your TaskList methods, it clearly explains what each of them achieves within the code!

shuyangk commented on src/main/java/dot/ui/Ui.java:
I like how you named your String variables within the Ui class.

shuyangk commented on src/main/java/dot/ui/Ui.java:
```diff
@@ -0,0 +1,109 @@
+
+    private static final String GOODBYE_MESSAGE = "Bye! DOnT forget to finish your tasks!\n";
+
+    private static final String HELP_MESSAGE = "Welcome to the help menu, DonT worry ^o^!\n"
+            + "<datetime> format for deadline/events: dd/MM/yyyy hhmm (e.g. 13/01/2020 1800)\n"
+            + "<date> format for whatsgoingon: dd/MM/yyyy (e.g. 13/01/2020)\n\n"
+            + "Commands:\n"
+            + "todo <description> - Add a todo task\n"
+            + "deadline <description> /by <deadline_description/datetime>\n"
+            + "event <description> /from <start_description/datetime> /to <start_description/datetime>\n"
+            + "list - List out all your tasks\n"
+            + "unmark <taskNo> - Unmark task based on no. on the list\n"
+            + "mark <taskNo> - Check off task based on no. on the list\n"
+            + "delete <taskNo> - Delete task based on no. on the list\n"
+            + "whatsgoingon <date> - See what deadlines/events are ongoing for given date\n"
+            + "find <query> - List out all tasks that fit <query>"
+            + "help - access this help menu\n"
+            + "bye - close Dot";
```
I like the inclusion of the HELP_MESSAGE to help users who might not be familiar with the commands!

lamchenghou replied:
Thanks for the review.

shuyangk reviewed (commented):
Overall I found your code easy to read. Code looks good, the naming conventions are followed, and the coding standard is also adhered to as well.

marioalvaro commented on src/main/java/dot/parser/Parser.java:
```diff
@@ -73,27 +73,35 @@ public static Command parseInputToCommand(String input,
                                               TaskList dotTaskList) throws DotException {
```
I think indentation for wrapped lines should be only 8 spaces

marioalvaro commented on src/main/java/dot/parser/Parser.java:
```diff
@@ -73,27 +73,35 @@ public static Command parseInputToCommand(String input,
                                               TaskList dotTaskList) throws DotException {
         if (input.strip().equals("bye")) {
```
For every if-else statement below, I could recommend using switch statement for better readability.

marioalvaro commented on src/main/java/dot/tasks/TaskList.java:
```diff
@@ -37,39 +38,12 @@ protected TaskList(int maxSize, Storage storage) {
+        // We will set maxSize to 2x of length of tasks, or maxSize, whichever is larger
         this.tasks = tasks;
-        this.maxSize = maxSize;
+        this.maxSize = Math.max(tasks.size() * 2, maxSize);
```
using this is not necessary as it does not lead to confusion

lamchenghou replied:
Thanks for the review! The fields here are being shadowed by constructor parameters.

marioalvaro commented on src/main/java/dot/commands/Command.java:
```diff
@@ -10,9 +12,11 @@ public abstract class Command {
      * @throws DotException passing the buck.
      */
-    public void execute() throws DotException {
+    public void execute(Consumer<String> handleDotOutput) throws DotException {
```
Nice approach using Consumer

marioalvaro commented on src/main/java/dot/ui/Ui.java:
```diff
@@ -1,91 +1,93 @@
 public class Ui {
 
-    private static final String HORIZONTAL_RULE = "_".repeat(80) + "\n";
+    private static final String HORIZONTAL_RULE = "_".repeat(38) + "\n";
```
Nice way to use .repeat() method. I did not know the function exist.

marioalvaro commented on src/main/java/dot/ui/Ui.java:
```diff
@@ -1,91 +1,93 @@
-    }
+    private static final String GOODBYE_MESSAGE = "Bye! DOnT forget to finish your tasks!\n";
+
+    private static final String HELP_MESSAGE = "Welcome to the help menu, DonT worry ^o^!\n"
```
Nice way to make new user understand

marioalvaro reviewed (commented):
Code is written nicely. But several parts of the code is lengthy and cramped together. Maybe try to break the code into methods for ease of reading. Other than that you may write @author [author] for every public class. 

Overall really nice!

lamchenghou committed:
Fix style issues based on reviews

lamchenghou committed:
Add assertions to some classes

There lacks a way to specify assumptions in the main code.
For instance, in Validation::getArgsIfValidDeadlineFormat,
we made the assumption that a regex is of a certain form
previously in comments.

Let's utilise assertions to ensure that the input is indeed
in the assume formed at that point in the code. We also assert
several conditions at other places in the codebase.

Assertion failures can indicate potential bugs in the code due
to an invalid assumption, which can then be corrected.

lamchenghou committed:
Separate lines of code based on what they do

lamchenghou committed:
Shorten code by extracting some methods

Some methods are too long, leading to low readability.

Let's extract some methods, mainly in hunks of similar purpose.

We extract similar code in order to make it logical to other
developers. It also helps to maintain readability through SLAP..

lamchenghou committed:
Add guard clauses to make happy paths prominent

The main code, should no exception be thrown, is nested
within conditionals, on the same level as exception throws.

Let's implement some guard clauses, in order to allow the
happy path to be at the top layer, improving readability.

Having a prominent happy paths is more desirable, not only
due to its readability, but in the case we want to add on
to this happy path in the future.

lamchenghou committed:
Extract variables in order to avoid complicated expressions

lamchenghou committed:
Use streams to read file for improvement in performance

lamchenghou committed:
Merge pull request #2 from lamchenghou/branch-A-Assertions

Add assertions to some classes

lamchenghou committed:
Merge branch 'master' into branch-A-CodeQuality

lamchenghou committed:
Merge pull request #3 from lamchenghou/branch-A-CodeQuality

Improve code quality

lamchenghou committed:
Merge branch 'master' into branch-A-Streams

lamchenghou committed:
Merge pull request #4 from lamchenghou/branch-A-Streams

Add streams to read file for improvement in performance

lamchenghou committed:
Add CI yml file

lamchenghou committed:
Add Undoable interface

lamchenghou committed:
Add UndoCommand and update commands to support undo

The undo feature will allow users to undo unintended
changes easily.

Let's add undo and isUndoable into the Command class
and add custom implementations for undo in each of
the appropriate subclasses. We also add an UndoCommand
to perform the undoing action.

Using polymorphism allows us to extend this undo feature
across multiple commands, with ease and readability.

lamchenghou committed:
Update tasks package classes and TaskErrors to support undo

lamchenghou committed:
Add undo functionality to Parser and MainWindow

lamchenghou committed:
Update help message to include undo

lamchenghou committed:
Update JUnit tests to follow undo

lamchenghou committed:
Fix DeleteCommand undo not saving to storage

lamchenghou committed:
Merge pull request #5 from lamchenghou/branch-C-Undo

Add undo feature

lamchenghou committed:
Set MainWindow to be resizable

lamchenghou committed:
Set font size and background to enlarge with window

The current window size might be too small for users
who prefer a large window.

We support resizability of the window. Components
scale appropriately with the window.

This allows a seamless user experience, and is
intuitive for all users.

lamchenghou committed:
Merge branch 'branch-A-BetterGui'

lamchenghou committed:
Fix assertions

lamchenghou committed:
Update user guide

lamchenghou committed:
Add Ui.png

lamchenghou committed:
Merge branch 'branch-A-UserGuide'

lamchenghou committed:
Fix user guide style issues

lamchenghou committed:
Remove horizontal rule in messages

lamchenghou committed:
Update Ui.png

lamchenghou committed:
Update User Guide

lamchenghou committed:
Fix Ui.png causing CI issues

lamchenghou committed:
Update README.md

lamchenghou committed:
Fix link in README

lamchenghou committed:
Fix coding standard issues

lamchenghou committed:
Update fxml files xmlns fx versions

lamchenghou committed:
Add images to pages

lamchenghou committed:
Add images to pages

lamchenghou committed:
Add line break
